bundles/auth/org.eclipse.smarthome.auth.oauth2client/src/main/java/org/eclipse/smarthome/auth/oauth2client/internal/OAuthClientServiceImpl.java

```diff
@@ -0,0 +1,356 @@
+
+        OAuthClientServiceImpl clientService = new OAuthClientServiceImpl(bundleContext, handle, tokenExpiresIn_Buffer);
+        clientService.storeHandler = storeHandler;
+
+        // Load parameters from Storage
+        clientService.persistedParams = storeHandler.loadPersistedParams(handle);
+        if (clientService.persistedParams == null) {
+            throw new OAuthException("Cannot get oauth provider parameters from handle:" + handle);
+        }
+        clientService.init();
+
+        return clientService;
+    }
+
+    /**
+     * It should only be used internally
+     *
+     * @param bundleContext
+     * @param storageHandler
+     * @param handle
+     * @return new instance of OAuthClientServiceImpl
+     *
+     * @throws
+     */
+    @SuppressWarnings("null")
+    static OAuthClientServiceImpl getInstance(@NonNull BundleContext bundleContext, @NonNull String handle,
+            @NonNull OAuthStoreHandler storeHandler, @NonNull String tokenUrl, @Nullable String authorizationUrl,
+            @NonNull String clientId, @Nullable String clientSecret, @Nullable String scope,
+            @Nullable Boolean supportsBasicAuth, int tokenExpiresIn_Buffer) {
+
+        OAuthClientServiceImpl clientService = new OAuthClientServiceImpl(bundleContext, handle, tokenExpiresIn_Buffer);
+        clientService.storeHandler = storeHandler;
+        clientService.persistedParams = new PersistedParams();
+        clientService.persistedParams.tokenUrl = tokenUrl;
+        clientService.persistedParams.authorizationUrl = authorizationUrl;
+        clientService.persistedParams.clientId = clientId;
+        clientService.persistedParams.clientSecret = clientSecret;
+        clientService.persistedParams.scope = scope;
+        clientService.persistedParams.supportsBasicAuth = supportsBasicAuth;
+
+        // this.persistedParams.tokenExpiresInBuffer
+        storeHandler.savePersistedParams(handle, clientService.persistedParams);
+
+        clientService.init();
+        return clientService;
+    }
+
+    @SuppressWarnings("null")
+    @Override
+    public String getAuthorizationUrl(@Nullable String redirectURI, @Nullable String scope, @Nullable String state) {
+        if (state == null) {
+            persistedParams.state = createNewState();
+        } else {
+            persistedParams.state = state;
+        }
+        persistedParams.scope = scope;
+
+        // keep it to check against redirectUri in #getAccessTokenResponseByAuthorizationCode
+        persistedParams.redirectUri = redirectURI;
+        return OAuthConnector.getAuthorizationUrl(this.persistedParams.authorizationUrl, this.persistedParams.clientId,
+                redirectURI, persistedParams.state, scope);
+    }
+
+    @Override
+    public String extractAuthCodeFromAuthResponse(@NonNull String redirectURLwithParams) throws OAuthException {
+        checkNotNull(redirectURLwithParams, "redirectURLwithParams must not be null");
```

Comment:
Since @NonNull has already been added, checkNotNull(..) precondition is not required